clojure-ts-mode.el

```diff
@@ -882,22 +884,87 @@ The returned value is expected to be the same as
+specification compatible with `clojure-mode', which will then be
+converted to a suitable `clojure-ts-mode' specification.
 
-The indentation rules for `clojure-ts-mode' are simpler than for
-`clojure-mode' so we only take the first integer N and produce `(:block
-N)' rule.  If an integer cannot be found, this function returns nil and
-the default rule is used."
+For example, (1 ((:defn)) nil) is converted to ((:block 1) (:inner 2))."
   (when (functionp clojure-ts-get-indent-function)
     (let ((spec (funcall clojure-ts-get-indent-function symbol-name)))
-      (if (consp spec)
-          `(:block ,(car spec))
-        (when (integerp spec)
-          `(:block ,spec))))))
+      (if (integerp spec)
+          (list (list :block spec))
+        (when (sequencep spec)
+          (thread-last spec
+                       (seq-map (lambda (el)
+                                  (cond
+                                   ((integerp el) (list :block el))
+                                   ((equal el :defn) (list :inner 0))
+                                   ((consp el) (clojure-ts--unnest-dynamic-spec el 0))
+                                   (t nil))))
+                       (seq-remove #'null)
+                       ;; Always put `:block' to the beginning.
+                       (seq-sort (lambda (spec1 _spec2)
+                                   (equal (car spec1) :block)))))))))
+
+(defun clojure-ts--find-semantic-rule (node parent current-depth)
+  "Returns a suitable indentation rule for NODE, considering the CURRENT-DEPTH.
+
+Attempts to find an indentation rule by examining the symbol name of the
+PARENT's first child.  If a rule is not found, it navigates up the
+syntax tree and recursively attempts to find a rule, incrementally
+increasing the CURRENT-DEPTH.  If a rule is not found upon reaching the
+root of the syntax tree, it returns nil.  A rule is considered a match
+only if the CURRENT-DEPTH matches the rule's required depth."
+  (let* ((first-child (clojure-ts--node-child-skip-metadata parent 0))
+         (symbol-name (clojure-ts--named-node-text first-child))
+         (idx (- (treesit-node-index node) 2)))
+    (if-let* ((rule-set (or (clojure-ts--dynamic-indent-for-symbol symbol-name)
+                            (alist-get symbol-name
+                                       (seq-union clojure-ts-semantic-indent-rules
+                                                  clojure-ts--semantic-indent-rules-defaults
+                                                  (lambda (e1 e2) (equal (car e1) (car e2))))
+                                       nil
+                                       nil
+                                       #'equal))))
+        (if (zerop current-depth)
+            (let ((rule (car rule-set)))
+              (if (equal (car rule) :block)
+                  rule
+                (let ((rule-depth (cadr rule))
+                      (rule-idx (caddr rule)))
```

Comment:
I'm not a big fan of `caddr` and the likes, because they test to confuse newcomers to Elisp a lot. I'd suggest going for destructuring functions (`pcase` or `cl-destructuring-bind`) instead or at least adding some examples in the comments of what you're aiming to match. There's also `pcase-let` you might find useful.

Reply:
Replaced with `pcase-let` (I left `car` in some places where it's in my opinion more appropriate).